tree/ntuple/v7/src/RNTuple.cxx

```diff
@@ -333,8 +337,14 @@ void ROOT::Experimental::RNTupleWriter::CommitCluster()
       field.Flush();
       field.CommitCluster();
    }
-   fSink->CommitCluster(fNEntries);
+   float nbytes = fSink->CommitCluster(fNEntries);
+   // Cap the compression factor at 1000 to prevent overflow of fMinUnzippedClusterSizeEst
+   float compressionFactor = std::min(1000.f, static_cast<float>(fUnzippedClusterSize) / nbytes);
+   fUnzippedClusterSizeEst =
```

Comment:
Using the avg. compression factor (instead of last cluster only) should probably be more accurate.

Reply:
> In reply to: [#8703 (comment)](https://github.com/root-project/root/pull/8703#issuecomment-920044106)
> 
> > You mean keeping the average of the last $x$ MB written per column?
> 
> (...)
> however conversely, an accurate estimate is "only important" for collection column, so maybe an hybrid solution is a workable compromise?

I think we have several viable strategies for a good estimator, each of them can be "tricked" by the input data. The problem you describe is not tied to collections, I think.  You might have, e.g., integer columns that are constantly zero until suddenly they are not anymore. In this case, even column-based estimators would guess wrong.

I added a discussion of possible alternatives to the tuning.md document. Unless shown to be a problem with real files, I think an advantage of the current strategy (look at the last cluster) is its very simple implementation, and that it works well on the test data sets we are using.

In any case, the cluster size is bounded by the maximum _uncompressed_ size, which is exact within the limits of the size of a single entry. It puts by default a 512MB limit to the size of a cluster in memory (and thus to its size on disk). If we ever reach this condition, it is certainly not ideal but probably still not catastrophic in many cases.  It would be also easy to diagnose and not too difficult to fix.

Reply:
@pcanal This comment ended up here, it was supposed to pick up both the notes from you and from Javier

Reply:
> The problem you describe is not tied to collections,

Yes but collections have the potential of amplifying the consequence of the missed estimate.

Reply:
> which is exact within the limits of the size of a single entry.

It can not be exact, there is also a way for the estimate "does the next entry brings us over the limit" to be wrong (because of the size variability of collections) and since the cluster "must" finish whatever entry it started is a collection of the last entry of the cluster is unusually large, you might have to go over the limit, isn't it?

> It would be also easy to diagnose

In practice, it is (too often) harder that one would expect.  For example, the failure mode might that some cluster are smaller than they should be (compression guess was too low) resulting in pervasive but not huge performance degradation (too many small reads from the file) that might becomes a "noticeable problem" only under "extreme condition" [e.g. thousands of HPC nodes reading from a high latency remote source].